kernels/volk/volk_32fc_x2_s32f_square_dist_scalar_mult_32f.h

```diff
@@ -79,69 +79,73 @@
+    diff = symbol - *points++;
+    *target++ = scalar * (lv_creal(diff) * lv_creal(diff) + lv_cimag(diff) * lv_cimag(diff));
+  }
+}
 
 
 #ifdef LV_HAVE_AVX2
 #include<immintrin.h>
+#include<volk/volk_avx2_intrinsics.h>
 
 static inline void
-volk_32fc_x2_s32f_square_dist_scalar_mult_32f_a_avx2(float* target, lv_32fc_t* src0, lv_32fc_t* points,
-                                                     float scalar, unsigned int num_points)
+volk_32fc_x2_s32f_square_dist_scalar_mult_32f_a_avx2(float* target, lv_32fc_t* src0, 
+                                                     lv_32fc_t* points, float scalar, 
+                                                     unsigned int num_points)
 {
   const unsigned int num_bytes = num_points*8;
-  __m128 xmm0, xmm9, xmm10, xmm11;
-  __m256 xmm1, xmm2, xmm3, xmm4, xmm5, xmm6, xmm7, xmm8;
-
-  lv_32fc_t diff;
-  memset(&diff, 0x0, 2*sizeof(float));
-
-  float sq_dist = 0.0;
-  int bound = num_bytes >> 6;
-  int leftovers0 = (num_bytes >> 5) & 1;
-  int leftovers1 = (num_bytes >> 4) & 1;
-  int leftovers2 = (num_bytes >> 3) & 1;
-  int i = 0;
-
-  __m256i idx = _mm256_set_epi32(7,6,3,2,5,4,1,0);
-  xmm1 = _mm256_setzero_ps();
-  xmm2 = _mm256_load_ps((float*)&points[0]);
-  xmm8 = _mm256_set1_ps(scalar);
-  xmm11 = _mm256_extractf128_ps(xmm8,1);
-  xmm0 = _mm_load_ps((float*)src0);
-  xmm0 = _mm_permute_ps(xmm0, 0b01000100);
-  xmm1 = _mm256_insertf128_ps(xmm1, xmm0, 0);
-  xmm1 = _mm256_insertf128_ps(xmm1, xmm0, 1);
-  xmm3 = _mm256_load_ps((float*)&points[4]);
-
-  for(; i < bound; ++i) {
-    xmm4 = _mm256_sub_ps(xmm1, xmm2);
-    xmm5 = _mm256_sub_ps(xmm1, xmm3);
+  __m128 xmm9, xmm10;
+  __m256 xmm4, xmm6;
+  __m256 xmm_points0, xmm_points1, xmm_result;
+
+  const unsigned int bound = num_bytes >> 6;
+  const unsigned int leftovers0 = (num_bytes >> 5) & 1;
+  const unsigned int leftovers1 = (num_bytes >> 4) & 1;
+  const unsigned int leftovers2 = (num_bytes >> 3) & 1;
+  
+  // load complex value into all parts of the register.
+  const __m256 xmm_symbol = _mm256_castpd_ps(_mm256_broadcast_sd((const double*)src0));
+  const __m128 xmm128_symbol = _mm256_extractf128_ps(xmm_symbol, 1);
+  
+  // Load scalar into all 8 parts of the register
+  const __m256 xmm_scalar = _mm256_broadcast_ss(&scalar);
+  const __m128 xmm128_scalar = _mm256_extractf128_ps(xmm_scalar, 1);
+
+  // Set permutation constant
+  const __m256i idx = _mm256_set_epi32(7,6,3,2,5,4,1,0);
+  
+  for(unsigned int i = 0; i < bound; ++i) {
+    xmm_points0 = _mm256_load_ps((float*)points);
+    xmm_points1 = _mm256_load_ps((float*)(points + 4));
     points += 8;
-    xmm6 = _mm256_mul_ps(xmm4, xmm4);
-    xmm7 = _mm256_mul_ps(xmm5, xmm5);
-
-    xmm2 = _mm256_load_ps((float*)&points[0]);
-
-    xmm4 = _mm256_hadd_ps(xmm6, xmm7);
-    xmm4 = _mm256_permutevar8x32_ps(xmm4, idx);
-
-    xmm3 = _mm256_load_ps((float*)&points[4]);
-
-    xmm4 = _mm256_mul_ps(xmm4, xmm8);
-
-    _mm256_store_ps(target, xmm4);
+    __VOLK_PREFETCH(points);
```

Comment:
I wonder whether this results in any speed gain (this is not the case on my Ivy Bridge machine). Given that the memory access pattern is extremely regular I think that the CPU will prefetch anyway. Skimming through AMD’s and Intel’s optimization guides indicates that meaningful usage of ``prefetch`` is not trivial. See also Agner Fog’s optimization guide [1, chapter 12.14]:

> Automatic data prefetching often works better than explicit data prefetching when
the data access pattern is sufficiently regular.
> […]
> A good deal of experimentation is often needed to find the optimal prefetching
strategy for a program that accesses data in an irregular manner.

[1] https://www.agner.org/optimize/optimizing_assembly.pdf

Reply:
The previous code structure basically tried to prefetch the next chunk of data as well. But in a dangerous way. Now, this should be more secure and the code should be more verbose. Also, this should clarify intention.

Reply:
> We had the "should we use `__VOLK_PREFETCH`" discussion in other threads as well. My conclusion was, to have it as long as it doesn't hurt.

Yes! Let's make reasonable efforts to not do it dangerously, and only do it if it doesn't hurt performance. The cache manager will auto-prefetch the data sometimes, but not others. Some compilers (& OSs for that matter) will support user-based prefetch, while others don't. We do our best to keep it safe & performance enhancing!